thorlcr/graph/thgraphmaster.cpp

```diff
@@ -1609,45 +1612,41 @@ void CJobMaster::broadcast(ICommunicator &comm, CMessageBuffer &msg, mptag_t mpt
         replyTag = queryJobChannel(0).queryMPServer().createReplyTag();
         msg.setReplyTag(replyTag);
     }
-    if (globals->getPropBool("@broadcastSendAsync", true)) // only here in case of problems/debugging.
+    auto f = [&](unsigned r)
     {
-        class CSendAsyncfor : public CAsyncFor
+        if (!aborting || comm.verifyConnection(r, (unsigned)-1, false))
         {
-            CMessageBuffer &msg;
-            mptag_t mptag;
-            unsigned timeout;
-            StringAttr errorMsg;
-            ICommunicator &comm;
-        public:
-            CSendAsyncfor(ICommunicator &_comm, CMessageBuffer &_msg, mptag_t _mptag, unsigned _timeout, const char *_errorMsg)
-                : comm(_comm), msg(_msg), mptag(_mptag), timeout(_timeout), errorMsg(_errorMsg)
-            {
-            }
-            void Do(unsigned i)
-            {
-                if (!comm.send(msg, i+1, mptag, timeout))
-                    throw createBCastException(i+1, errorMsg);
-            }
-        } afor(comm, msg, mptag, timeout, errorMsg);
+            if (!comm.send(msg, r, mptag, timeout))
+                throw createBCastException(r, errorMsg);
+        }
+    };
+    std::vector<std::future<void>> sendResultsFuture;
+    Owned<IException> sendExcept;
+    for (unsigned dstrank=1; dstrank<=groupSizeExcludingMaster; ++dstrank)
+        sendResultsFuture.push_back(std::async(f, dstrank));
+    for (auto & sendResultFuture: sendResultsFuture)
+    {
         try
         {
-            afor.For(groupSizeExcludingMaster, groupSizeExcludingMaster);
+            sendResultFuture.get();
         }
         catch (IException *e)
         {
-            EXCLOG(e, "broadcastSendAsync");
-            abort(e);
-            throw;
+            if (!sendExcept)
+                sendExcept.setown(e);
+            else
+                e->Release();
         }
     }
-    else if (!comm.send(msg, RANK_ALL_OTHER, mptag, timeout))
+    if (sendExcept)
     {
-        Owned<IException> e = createBCastException(0, errorMsg);
-        EXCLOG(e, NULL);
-        abort(e);
-        throw e.getClear();
+        EXCLOG(sendExcept, "broadcastSendAsync");
+        abort(sendExcept);
+        throw sendExcept.getClear();
     }
+
     if (sendOnly) return;
+    groupSizeExcludingMaster = sendResultsFuture.size(); // actual number of broadcast recipients
```

Comment:
sendResultsFuture.size is always going to equal the original value of groupSizeExcludingMaster afaics ?

I think you need to count how many actually sent, i.e. if aborting, how many of the threads passed verifyConnection.
Could do that by returning a boolean from the async function.